feat: Add health monitoring server with comprehensive system checks, closes #49

ricardo-perello opened the pull request:
# feat: Add health monitoring server with comprehensive system checks, closes #49

## Overview
This PR adds a comprehensive health monitoring server to rindexer that provides real-time monitoring of system components and services.

## Features Added

### Health Server Module
- **New health server module** with HTTP endpoint at `/health`
- **Real-time monitoring** of database, indexing, and sync status
- **Health configuration** in manifest with default port 8080
- **Integration** into all CLI start commands (indexer, graphql, all)

### Health Checks Implemented
- **Database connectivity and health** - Verifies PostgreSQL connection status
- **Indexing service status** - Monitors active indexing tasks and service state
- **Data synchronization status** - Checks PostgreSQL tables or CSV files for data presence
- **Overall system health** - Provides timestamped health status with service breakdown

### Dependencies Added
- **axum** - HTTP server framework for the health endpoint

### Configuration
- **Health server enabled by default** in new project templates
- **Backward compatible** - existing manifests without health configuration use sensible defaults
- **Parallel operation** - runs alongside existing services without impacting core indexing functionality

## Technical Details

### Health Endpoint Response
```json
{
  "status": "healthy",
  "timestamp": "2025-09-04T02:48:40.370651Z",
  "services": {
    "database": "healthy",
    "indexing": "healthy", 
    "sync": "healthy"
  },
  "indexing": {
    "active_tasks": 1,
    "is_running": true
  }
}
```

### Health Server Architecture
- **Isolated postgres client** - Health server creates its own database connection for monitoring without interfering with main indexer operations
- **Non-blocking operation** - Runs in separate tokio task alongside GraphQL server
- **Graceful error handling** - Continues operation even if health checks fail

## Bug Fixes
- **Fixed database race condition** - Resolved issue where health server was interfering with main indexer's database schema setup
- **Proper client isolation** - Health server now only creates monitoring client without calling `setup_postgres()`

## Examples Updated
- Updated examples to include health server configuration
- Health server runs on port 8080 by default
- Accessible at `http://localhost:8080/health`

## Testing
- ✅ Health server starts successfully alongside indexer
- ✅ Database connectivity monitoring works
- ✅ Indexing status monitoring works  
- ✅ No interference with core indexing functionality
- ✅ Backward compatibility maintained

## Closes
Health monitoring requirements #49

joshstevens19 commented on cli/src/commands/start.rs:
```diff
@@ -182,6 +182,10 @@ pub async fn start(
                     },
+                    health_details: HealthOverrideSettings {
+                        enabled: true,
+                        override_port: None,
```
you need to cross check wherever it does it now that port sent for the graphql vs health never matches else bail out before was not an issue before only 1 port but now 2 overrides have to handle that edge case

joshstevens19 commented on cli/src/commands/start.rs:
```diff
@@ -182,6 +182,10 @@ pub async fn start(
                         override_port: None,
                     },
+                    health_details: HealthOverrideSettings {
+                        enabled: true,
```
any reason we even want anyone to ever turn it off? feels like something making the yaml file more clunky i would probs add this to the global one https://rindexer.xyz/docs/start-building/yaml-config/global as a single thing aka health_override_port: and then make it always on

joshstevens19 commented on core/src/health.rs:
```diff
@@ -0,0 +1,214 @@
+        },
+    };
+
+    // Check database connection if PostgreSQL is enabled
```
remove these kind of comments as not super useful we can tell by the function name its doing that

joshstevens19 commented on core/src/health.rs:
```diff
@@ -0,0 +1,214 @@
+    }
+}
+
+async fn health_handler(State(state): State<HealthServerState>) -> Result<Json<HealthStatus>, StatusCode> {
```
this function should split up and then build the object with the function calls its to long hard to read and way to many if elses

```
HealthStatus {
        status: "healthy".to_string(),
        timestamp: chrono::Utc::now().to_rfc3339(),
        services: HealthServices {
            database: database_health_check(),
            ...
        },
        indexing: IndexingStatus {
            active_tasks: active_indexing_count(),
            is_running: is_running(),
        },
    }
   ```

joshstevens19 commented on core/src/health.rs:
```diff
@@ -0,0 +1,214 @@
+
+async fn health_handler(State(state): State<HealthServerState>) -> Result<Json<HealthStatus>, StatusCode> {
+    let mut health_status = HealthStatus {
+        status: "healthy".to_string(),
```
should be an enun

joshstevens19 commented on core/src/health.rs:
```diff
@@ -0,0 +1,214 @@
+            Some(client) => {
+                match client.query_one("SELECT 1", &[]).await {
+                    Ok(_) => {
+                        health_status.services.database = "healthy".to_string();
```
enums and should be in its own method and make it cleaner we should try not to mut the health object and build it in 1 - same comments for most this file

joshstevens19 commented on core/src/health.rs:
```diff
@@ -0,0 +1,214 @@
+        match &state.postgres_client {
+            Some(client) => {
+                match client.query_one(
+                    "SELECT COUNT(*) FROM information_schema.tables WHERE table_schema = 'public' AND table_name LIKE '%_events'",
```
count(*) is expensive you should just check if it exists and do the SELECT 1 FROM ... LIMIT 1

joshstevens19 commented on core/src/health.rs:
```diff
@@ -0,0 +1,214 @@
+    if state.manifest.storage.postgres_enabled() {
+        match &state.postgres_client {
+            Some(client) => {
+                match client.query_one(
```
exists

joshstevens19 commented on core/src/health.rs:
```diff
@@ -0,0 +1,214 @@
+        }
+    } else {
+        // For CSV storage, check if the output directory exists and has files
+        if state.manifest.storage.csv_enabled() {
```
method to nested shouldnt need to write it like this once you split this function up and you see the after you understand what i mean

joshstevens19 commented on core/src/manifest/health.rs:
````diff
@@ -0,0 +1,60 @@
+/// ```
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct HealthSettings {
````
no need to be in manifest as per my above comment

joshstevens19 commented on core/src/start.rs:
```diff
@@ -153,6 +160,37 @@ pub async fn start_rindexer(details: StartDetails<'_>) -> Result<(), StartRindex
+                Some(tokio::spawn(async move {
+                    // Create postgres client for health server if postgres is enabled
+                    // Note: We only create a client for health checks, not setup the database schema
+                    let postgres_client = if manifest_clone.storage.postgres_enabled() {
```
there is a initialize_database which does exactly this

joshstevens19 reviewed (changes requested):
good PR just need some general cleaning up / making it more readable and a few other changes

ricardo-perello commented:
Here's the PR text with a small explanation about the SQL query:

---

## PR Review Response - All Feedback Addressed

Hi @joshstevens19! Thanks for the detailed feedback. I've addressed all the issues you highlighted. Here's what I've implemented:

### **Port Conflict Validation**
**Issue**: Need to cross-check GraphQL vs health server ports to prevent conflicts
**Solution**: Added port conflict validation in `start.rs` that checks both ports before starting either service and bails out with a clear error message if they match.

### **Simplified Configuration Architecture** 
**Issue**: Health server configuration was clunky with separate manifest section
**Solution**: 
- Moved health configuration to `global.health_override_port` (always enabled by default)
- Removed separate `HealthSettings` struct and `health.rs` manifest file
- Health server now runs on port 8080 by default, configurable via global settings
- Much cleaner and more intuitive configuration

### **Code Quality Improvements**
**Issue**: Large `health_handler` function was hard to read
**Solution**: Refactored into smaller, focused functions:
- `build_health_status()` - constructs the response object
- `check_database_health()` - database connectivity check
- `check_indexing_health()` - indexing service status
- `check_sync_health()` - data synchronization status
- `determine_overall_status()` - calculates overall health

### **Performance Optimizations**
**Issue**: Database queries were inefficient
**Solution**: 
- Replaced `COUNT(*)` queries with `query_one_or_none()` for better performance
- Uses `EXISTS` logic instead of counting all rows
- Health server now uses `initialize_database()` function for proper client setup

**Database Sync Health Check**: The sync health check uses an optimized SQL query that checks for the existence of user data tables while filtering out system tables:
```sql
SELECT 1 FROM information_schema.tables 
WHERE table_schema NOT IN ('information_schema', 'pg_catalog', 'rindexer_internal') 
AND table_name NOT LIKE 'latest_block' 
AND table_name NOT LIKE '%_last_known_%' 
AND table_name NOT LIKE '%_last_run_%' 
LIMIT 1
```
This efficiently determines if any meaningful user data exists (returns 'healthy') or if the indexer hasn't synced any events yet (returns 'no_data').

### **Better Integration**
**Issue**: Health server wasn't properly integrated with existing systems
**Solution**:
- Health server follows indexer lifecycle (only runs when indexer is running)
- Uses existing `initialize_database()` function instead of custom postgres setup
- Proper error handling and graceful degradation
- No interference with core indexing functionality

The health server is now production-ready with a clean, maintainable codebase that follows all the feedback you provided. All the architectural concerns have been addressed while maintaining the core functionality.

joshstevens19 commented on documentation/docs/pages/docs/start-building/running.mdx:
```diff
@@ -37,6 +37,151 @@ If you change your contract ABIs or want to start fresh you can use the [delete]
 You can also use the [drop_each_run](/docs/start-building/yaml-config/storage#drop_each_run) option in the YAML configuration file to drop all the data for the indexer before starting.
 :::
 
+## Health Monitoring
```
this needs moving to its own file or something as running is easy and quick and this takes up a bit to much room and added context
